lib/CPathfinder.cpp

```diff
@@ -79,96 +86,118 @@ void CPathfinder::calculatePaths()
 	//logGlobal->infoStream() << boost::format("Calculating paths for hero %s (adress  %d) of player %d") % hero->name % hero % hero->tempOwner;
 
 	//initial tile - set cost on 0 and add to the queue
-	CGPathNode &initialNode = *getNode(out.hpos);
-	initialNode.turns = 0;
-	initialNode.moveRemains = hero->movement;
-	mq.push_back(&initialNode);
+	CGPathNode * initialNode = out.getNode(out.hpos, hero->boat ? ELayer::SAIL : ELayer::LAND);
+	initialNode->turns = 0;
+	initialNode->moveRemains = hero->movement;
+	pq.push(initialNode);
 
-	while(!mq.empty())
+	while(!pq.empty())
 	{
```

Comment:
Just a notice about performance:
Each CPU cycle inside this (outer) loop consume 0.12% of overall time.